src/lib/prepareParams.js

```diff
@@ -0,0 +1,134 @@
+};
+
+function digestParameters(parameters) {
+  return parameters.reduce((prev, param) => {
+    if ('$ref' in param || 'allOf' in param || 'anyOf' in param || 'oneOf' in param) {
+      throw new Error(`The OpenAPI document for this operation wasn't dereferenced before processing.`);
+    } else if (param.name in prev) {
+      throw new Error(`The operation you are using has the same parameter, ${param.name}, spread across multiple entry points. We unfortunately can't handle this right now.`);
+    }
+
+    return Object.assign(prev, { [param.name]: param });
+  }, {});
+}
+
+// https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_isempty
+function isEmpty(obj) {
+  return [Object, Array].includes((obj || {}).constructor) && !Object.entries((obj || {})).length;
+}
+
+module.exports = function (operation, body, metadata) {
+  // If no data was supplied, just return immediately.
+  if (isEmpty(body) && isEmpty(metadata)) {
+    return {};
+  }
+
+  const params = {};
+  let shouldDigestParams = false;
+  const contentType = operation.getContentType();
+
+  if (Array.isArray(body)) {
+    // If the body param is an array, then it's absolutely a body and not something we need to do analysis against.
+    params.body = body;
+
+    if (typeof metadata !== 'undefined') {
+      shouldDigestParams = true;
+    }
+  } else if (typeof metadata === 'undefined') {
+    // No metadata was explicitly defined so we need to analyze the body to determine if it should actually be treated
+    // as metadata.
+    shouldDigestParams = true;
+  } else {
+    // Body and metadata were both supplied.
+    params.body = body;
+    shouldDigestParams = true;
+  }
+
+  let digested = {};
+  let hasDigestedParams = false;
+  if (shouldDigestParams) {
+    // @todo `operation.parameters` should also pull in common params (this does not happen automatically when dereffing!)
+    if ('parameters' in operation) {
+      digested = digestParameters(operation.parameters);
+      hasDigestedParams = Object.keys(digested).length;
+    }
+  }
+
+  // No metadata was explicitly defined so we need to analyze the supplied, and we haven't already set a body then we
+  // need to analyze the supplied body to see if it should actually be metadata. If not, then we can just treat it as a
+  // body and pass it along.
+  if (!('body' in params) && typeof metadata === 'undefined') {
+    if (!hasDigestedParams) {
+      // No parameters were able to be digested, so we just have to assume that what the user supplied was for a body.
+      // This might lead to unwanted false positives if an OAS isn't accurate, but short of throwing an error there
+      // isn't anything we can really do about it.
+      params.body = body;
+    } else {
+      const intersection = Object.keys(body).filter(value => Object.keys(digested).includes(value)).length;
+      if (intersection && (intersection / Object.keys(body).length) > 0.25) {
```

Comment:
Would love to see some tests around this logic.

Reply:
Done!